### ISSUE TITLE
Improve binary/multi_class likelihood precision

### ISSUE THREAD

**Issue by fehiepsi** [pyro-ppl/pyro#1996]
Currently, we use `link_function` as a map from `logits` to `probs`, and feed `probs` into Binary or Categorical distributions. We can improve the performance of these likelihoods by use `logits` directly in these distributions.
